src/handlers/no_merges.rs

```diff
@@ -0,0 +1,141 @@
+//! Purpose: When opening a PR, or pushing new changes, check for merge commits
+//! and notify the user of our no-merge policy.
+
+use crate::{
+    config::NoMergesConfig,
+    db::issue_data::IssueData,
+    github::{IssuesAction, IssuesEvent},
+    handlers::Context,
+};
+use anyhow::Context as _;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+use std::fmt::Write;
+use tracing as log;
+
+const NO_MERGES_KEY: &str = "no_merges";
+
+pub(super) struct NoMergesInput {
+    /// Hashes of merge commits in the pull request.
+    merge_commits: HashSet<String>,
+    /// Hash of the commit before the first merge commit.
+    before_first_merge_commit: String,
+}
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+struct NoMergesState {
+    /// Hashes of merge commits that have already been mentioned by triagebot in a comment.
+    mentioned_merge_commits: HashSet<String>,
+}
+
+pub(super) async fn parse_input(
+    ctx: &Context,
+    event: &IssuesEvent,
+    config: Option<&NoMergesConfig>,
+) -> Result<Option<NoMergesInput>, String> {
+    if !matches!(
+        event.action,
+        IssuesAction::Opened | IssuesAction::Synchronize | IssuesAction::ReadyForReview
+    ) {
+        return Ok(None);
+    }
+
+    // Require an empty configuration block to enable no-merges notifications.
+    if config.is_none() {
+        return Ok(None);
+    }
+
+    // Don't ping on rollups or draft PRs.
+    if event.issue.title.starts_with("Rollup of") || event.issue.draft {
+        return Ok(None);
+    }
+
+    let mut merge_commits = HashSet::new();
+    let mut before_first_merge_commit = None;
+    let commits = event
+        .issue
+        .commits(&ctx.github)
+        .await
+        .map_err(|e| {
+            log::error!("failed to fetch commits: {:?}", e);
+        })
+        .unwrap_or_default();
+    for commit in commits {
+        if commit.parents.len() > 1 {
+            merge_commits.insert(commit.sha.clone());
+        }
+
+        if merge_commits.is_empty() {
+            before_first_merge_commit = Some(commit.sha);
```

Comment:
I'm not sure I follow this logic. Is there really an ordering guarantee on the commits returned? I would expect the answer is no, or that the answer is "yes" (but that guarantee is by commit date, which is useless here).

I think github returns the base commit via the get pull request API (https://docs.github.com/en/rest/pulls/pulls#get-a-pull-request), which might be a better mechanism for this.

Reply:
I've just removed this and always suggest `master` - my thinking was that if the commits were `A -> B -> C -> D (Merge)` then it would suggest the sha for the parent commit in the pull request of the merge commit - `C` here.